File: examples/rust/main/button.rs
use warduino::*;

static BUTTON : u32 = 25;
static LED    : u32 = 26;

fn callback(_topic: &str, _payload: &str, _length: u32) {
    let val = digital_read(LED);
    //print(format!("Switch LED to {}\n", (val + 1) % 2).as_bytes());
    if val == HIGH {
        digital_write(LED, LOW);
    } else {
        digital_write(LED, HIGH);
    }
}

#[no_mangle]
pub fn main() {
    pin_mode(BUTTON, INPUT);
    pin_mode(LED, OUTPUT);

    sub_interrupt(BUTTON, callback, FALLING);

    loop {
        //delay(1);
    }
}

File: examples/assemblyscript/main/button.ts
import {
    delay,
    digitalRead,
    digitalWrite,
    InterruptMode,
    interruptOn,
    pinMode,
    PinMode,
    PinVoltage
} from "as-warduino/assembly";

const button = 25;
const led = 26;

function invert(voltage: PinVoltage): PinVoltage {
    switch (voltage) {
        case PinVoltage.LOW:
            return PinVoltage.HIGH;
        case PinVoltage.HIGH:
        default:
            return PinVoltage.LOW;
    }
}

function toggleLED(_topic: string, _payload: string): void {
    // Get current status of LED
    let status = digitalRead(led);
    // Toggle LED
    digitalWrite(led, invert(status));
}

export function main(): void {
    pinMode(button, PinMode.INPUT);
    pinMode(led, PinMode.OUTPUT);

    interruptOn(button, InterruptMode.FALLING, toggleLED);

    while (true) {
        delay(1000);
    }
}
